globus_sdk/config/environments.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import typing
+
+log = logging.getLogger(__name__)
+
+
+class EnvConfig:
+    envname: str
+    domain: str
+    no_dotapi: typing.List[str] = ["auth"]
+
+    # this same dict is inherited (and therefore shared!) by all subclasses
+    _registry: typing.Dict[str, typing.Type["EnvConfig"]] = {}
+
+    # this is an easier hook to use than metaclass definition -- register every subclass
+    # in this dict automatically
+    #
+    # as a result, anyone can define
+    #
+    #       class BetaEnv(EnvConfig):
+    #           domain = "beta.foo.bar.example.com"
+    #           envname = "beta"
+    #
+    # and retrieve it with get_config_by_name("beta")
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._registry[cls.envname] = cls
+
+    @classmethod
+    def get_service_url(cls, service):
+        # you can override any name with a config attribute
+        service_url_attr = f"{service}_url"
+        if hasattr(cls, service_url_attr):
+            return getattr(cls, service_url_attr)
+
+        # the typical pattern for a service hostname is X.api.Y
+        # X=transfer, Y=preview.globus.org => transfer.api.preview.globus.org
+        # check `no_dotapi` for services which don't have `.api` in their names
+        if service in cls.no_dotapi:
+            return f"https://{service}.{cls.domain}/"
+        return f"https://{service}.api.{cls.domain}/"
```

Comment:
I had a lot of uncertainty about how magical to make this behavior. How much do we do here, vs how much do we just leave for subclasses to define for themselves. I think this was about the right amount of "magic" to include, and it can always be overridden with explicit `transfer_url = ...`, etc.
I'm open to discussion and changes on this if it's too unsightly, however.

One thing that I do kind of like is that this makes the SDK answerable for new services without any changes here. Ask for `get_service_url("groups")`, and a predictable, probably correct, thing happens.